DigitalLearningSolutions.Data/Services/CourseContentService.cs

```diff
@@ -80,18 +82,67 @@ GROUP BY
             ).FirstOrDefault();
         }
 
+        public bool DoesProgressExist(int candidateId, int customisationId)
+        {
+            return connection.Query<int>(
+                @"SELECT ProgressId
+                        FROM Progress
+                        WHERE CandidateID = @candidateId
+                          AND CustomisationID = @customisationId
+                          AND SystemRefreshed = 0
+                          AND RemovedDate IS NULL",
+                new { candidateId, customisationId }
+            ).Any();
+        }
+
         public int GetProgressId(int candidateId, int customisationId)
         {
-            // TODO HEEDLS-202: change QueryFirstOrDefault to creating progress record if not found
             return connection.QueryFirstOrDefault<int>(
                 @"SELECT ProgressId
                         FROM Progress
                         WHERE CandidateID = @candidateId
-                          AND CustomisationID = @customisationId",
+                          AND CustomisationID = @customisationId
+                          AND SystemRefreshed = 0
+                          AND RemovedDate IS NULL",
                 new { candidateId, customisationId }
             );
         }
 
+        public void InsertNewProgress(int candidateId, int customisationId, int centreId)
+        {
+            var errorCode = connection.QueryFirst<int>(
+                @"uspCreateProgressRecord_V3",
+                new
+                {
+                    candidateId,
+                    customisationId,
+                    centreId,
+                    EnrollmentMethodID = 1,
+                    EnrolledByAdminID = 0
+                },
+                commandType: CommandType.StoredProcedure
+            );
+
+            switch (errorCode)
+            {
+                case 1:
+                    logger.LogWarning(
+                        "Not enrolled candidate on course as progress already exists. " +
+                        $"Candidate id: {candidateId}, customisation id: {customisationId}, centreId{centreId}");
+                    break;
+                case 100:
+                    logger.LogWarning(
+                        "Not enrolled candidate on course as customisation id doesn't match centre id. " +
+                        $"Candidate id: {candidateId}, customisation id: {customisationId}, centreId{centreId}");
+                    break;
+                case 101:
+                    logger.LogWarning(
+                        "Not enrolled candidate on course as candidate id doesn't match centre id. " +
+                        $"Candidate id: {candidateId}, customisation id: {customisationId}, centreId{centreId}");
+                    break;

```

Comment:
We probably want a default case as well, I'd expect the IDE to warn you about that?

Reply:
The IDE doesn't give any warnings, we don't expect to obtain any error codes that aren't listed so I didn't think we needed to define it. Should I include it just in case?

Reply:
I'm pretty sure there'll be a small green squigly line under the switch keyword, at least there is when I try it. Yes you should have a default case because the stored procedure may be modified 🙂